ProofWidgets/Component/SelectInsertPanel.lean

```diff
@@ -0,0 +1,113 @@
+
+This file builds infrastructure to define tactics allowing users to select things in the info
+view and generate a tactic call. For instance one can use this infrastructure to build
+a `conv?` tactic that allows to select a subterm of the goal and, after clicking on some link
+below the tactic state, replace the `conv?` call by a `conv => enter [...]` call zooming on the
+selected subterm.
+
+The only public part of this file is the `mkSelectInsertTactic` macro. Other declarations
+are internal details of the implementation of this macro. The corresponding JaveScript code,
+which is in `widget/src/selectInsertPanel.tsx`, is also an implemantation detail.
+-/
+
+/-- Parameter for a widget callback that will insert a tactic call after selecting things in
+the tactic state. -/
+structure SelectInsertParams where
+  /-- The current cursor position. -/
+  cursorPos : Lsp.Position
+  /-- The current context information. -/
+  ctx : WithRpcRef Elab.ContextInfo
+  /-- The locations selected in the tactic state. -/
+  selectedLocations : Array SubExpr.GoalsLocation
+  /-- The range in the source document where the result will be inserted. -/
+  replaceRange : Lsp.Range
+  deriving RpcEncodable
+
+/-- Response for a widget callback that will insert a tactic call after selecting things in
+the tactic state. -/
+structure SelectInsertResponse where
+  /-- The tactic call to be displayed in the panel. -/
+  content : String
+  /-- The edit to perform on the file. -/
+  edit : Lsp.WorkspaceEdit
+  /-- Where to place the cursor after the edit. It is always in the same file. -/
+  newCursorPos : Lsp.Position
+  deriving FromJson, ToJson
+
+/-- Create the function that will actually handle the call from the widget. -/
+def mkSelectInsertCommand
+    (mkCmdStr : (pos : Array Lean.SubExpr.GoalsLocation) → (goalType : Expr) → MetaM String) :
+    SelectInsertParams → RequestM (RequestTask (Option SelectInsertResponse)) :=
+  fun ⟨cursorPos, ⟨ctx⟩, selectedLocations, replaceRange⟩ ↦
+    RequestM.withWaitFindSnapAtPos cursorPos fun _ => do
+      let some firstLoc := selectedLocations[0]? | return none
+      let doc ← RequestM.readDoc
+      ctx.runMetaM {} do
+        let md ← firstLoc.mvarId.getDecl
+        let lctx := md.lctx |>.sanitizeNames.run' {options := (← getOptions)}
+        Meta.withLCtx lctx md.localInstances do
+          let insertedCode ← mkCmdStr selectedLocations md.type
+          let textDocumentEdit : Lsp.TextDocumentEdit := {
+            textDocument := { uri := doc.meta.uri, version? := doc.meta.version },
+            edits        := #[{range := replaceRange, newText := insertedCode }] }
+          return some { content := insertedCode,
+                        edit := Lsp.WorkspaceEdit.ofTextDocumentEdit textDocumentEdit,
+                        /- CHECKME: the new cursor position computation below looks very naive.
+                           Will it work with non-ascii insertedCode? -/
```

Comment:
Let me highlight there is a question here. @Vtec234

Reply:
Fixed during refactor, though note the cursor may still end up mispositioned due to leanprover/vscode-lean4#225.